cmake/extensions.cmake

```diff
@@ -371,6 +371,93 @@ function(generate_inc_file_for_target
   generate_inc_file_for_gen_target(${target} ${source_file} ${generated_file} ${generated_target_name} ${ARGN})
 endfunction()
 
+function(target_sources_codegen target)
+  foreach(arg ${ARGN})
+    # Generated file must be generated to the current binary directory.
+    # Otherwise this would trigger CMake issue #14633:
+    # https://gitlab.kitware.com/cmake/cmake/issues/14633
+    if(IS_ABSOLUTE ${arg})
+      set(template_file ${arg})
+      get_filename_component(generated_file_name ${arg} NAME)
+      set(generated_file ${CMAKE_CURRENT_BINARY_DIR}/${generated_file_name})
+    else()
+      set(template_file ${CMAKE_CURRENT_SOURCE_DIR}/${arg})
+      set(generated_file ${CMAKE_CURRENT_BINARY_DIR}/${arg})
+    endif()
+
+    if(IS_DIRECTORY ${template_file})
+      message(FATAL_ERROR "target_sources_codegen() was called on a directory")
+    endif()
+
+    # Generate file from template
+    message(STATUS "Will generate ${generated_file}")
+    message(STATUS "         from ${template_file}")
+    add_custom_command(
+      OUTPUT ${generated_file}
+      DEPENDS ${template_file}
+      COMMENT "Generating ${generated_file}"
+              "      from ${template_file}"
+      COMMAND
+      ${PYTHON_EXECUTABLE}
+      ${ZEPHYR_BASE}/scripts/gen_code.py
+      -D "PROJECT_NAME=${PROJECT_NAME}"
+      -D "PROJECT_SOURCE_DIR=${PROJECT_SOURCE_DIR}"
+      -D "PROJECT_BINARY_DIR=${PROJECT_BINARY_DIR}"
+      -D "CMAKE_SOURCE_DIR=${CMAKE_SOURCE_DIR}"
+      -D "CMAKE_BINARY_DIR=${CMAKE_BINARY_DIR}"
+      -D "CMAKE_CURRENT_SOURCE_DIR=${CMAKE_CURRENT_SOURCE_DIR}"
+      -D "CMAKE_CURRENT_BINARY_DIR=${CMAKE_CURRENT_BINARY_DIR}"
+      -D "CMAKE_CURRENT_LIST_DIR=${CMAKE_CURRENT_LIST_DIR}"
+      -D "CMAKE_FILES_DIRECTORY=${CMAKE_FILES_DIRECTORY}"
```

Comment:
Users must be able to specify additional CMake variables that 
should be imported into the codegen namespace.

E.g. to generate offsets.h, it is necessary to know the path of offsets.o.

Also, we might want to cut down on the default list to minimize scope. But it is not clear to me
whether this should be done or not.

Reply:
Can´t you use the inline code to get this information?

Reply:
I can't think of a way for the inline code to deduce this, so no, not without duplicating this information, which I would rather not do.

The build system knows where the file is and should pass this information on to the script. When things are organized like this the scope is minimized and it is easier to understand what the script is doing. (Because we can see exactly what kind of information is fed to it).